### ISSUE TITLE
Support generic equality/equivalent operators in Translator

### ISSUE THREAD

**Issue by duncand** [cqframework/clinical_quality_language#261]
See the following example CQL code:

```
library TestGenericEqual
define Test: 24 = 'hello'
define Test: Equal(24,'hello')
```

This currently fails to translate citing no matching signature ```Equal(Integer, String)``` but each define should be resolving to an ```Equal(Any, Any)```.

CQL spec section "2.5.1 Comparison Operators" says: "the equality and inequality operators can be used on any type of value within CQL"

CQL spec section "3.4.1 System-Defined Types" says: "The maximal supertype is System.Any. All other types derive from System.Any, meaning that any value is of some type, and also ultimately of type System.Any."

Given that the CQL type system features sub-typing and Any includes all values of the type system, it is logical for Equal(Any,Any) to exist and it has well-defined semantics:  Equal(x,y) returns true if and only if x and y are the same value, full stop.  When x and y are members of mutually disjoint types, the result of Equal(x,y) is simply false.

And so, Equal(x,y) should ALWAYS succeed in translating, regardless of the types of x and y.

Everything said above also applies to the operator Equivalent(Any,Any) aka "~".

**Comment by cmoesel** [comment]
This is an interesting topic.  To play devil's advocate -- what is the advantage of allowing a comparison that is _known to always be false_?  It seems to me that this would indicate a bug in the logic -- and perhaps is better surfaced at compile time than at run time.

Can you point to a case where the result may actually be true or false, but the compiler chokes on the `Equals` expression?

(Of course, to play devil's advocate to my own devil's advocate, the compiler _does_ allow `24 = 24`, which is known at compile time to always be true, and is also likely a bug in logic.)

**Comment by duncand** [comment]
Trying to evaluate an expression with no free variables is not a logical error due to having no free variables and it should not fail to translate solely for that reason.  One can perform constant expression folding at compile time however if the goal is performance or conciseness, though the Translator itself shouldn't do this by default in order to preserve the user's original logical structure.

But a constant expression isn't the main use case here, rather it is a simplified example.

Normally one or both of the arguments would be an arbitrary expression or parameter of type Any, typically when one is wanting to define a generic function, like this for example:

```
define private function foo(x Any, y Any, message String):
    bar(x ~ y, message)
```

It is a core tenet of good programming language design that if you have a value expression that may take the name of a variable or parameter or variable sub-expression in a particular place, you can also use a literal or constant expression in the same place.  Its a basic principle of orthogonality and of least surprise.

**Comment by brynrhodes** [comment]
Can you help me understand what you mean by the term _free variable_ in this context and how it applies to this case?

@cmoesel is suggesting, quite reasonably, that attempting to compare values of different types is a logical error that the compiler can help the author detect. That's the whole point of the type system in the first place, and exactly my concern with introducing a generic equal that would circumvent that capability. Yes, it would help with writing generic functions, but the ability to write generic functions is not the target use case for this language.

**Comment by duncand** [comment]
My term "free variable" was meant in a logic/math sense as an input to an expression which could vary for each evaluation; this was the opposite of a constant; for example, in "42 + x * 3 - y" the x and y are the free variables.  In the narrower sense, a free variable is an input not known at parse or compile time, because it is a parameter to a function containing the expression that can change per function call; in the wider sense, it is ultimately derived from a library parameter or external resource known only at runtime.

The point of the type system is to formally define how we classify and operate on data such that operations considered valid have well-defined behaviours and operations considered invalid don't execute or compile.

There is a key difference between an action that isn't logically valid and an action that the author is performing by mistake, and the type system should be distinguishing these.  The first should produce a fatal error that stops compilation or execution while the latter should merely display a warning but proceed with compilation anyway, as the author may have done it on purpose.

Really, "errors" and "warnings" are completely distinct classes, and there are a lot of things which could be made to receive warnings in the future that currently compile as the compiler becomes smarter about guessing what may have been author mistakes.

For example, per @cmoesel comment, the compiler allows "25 = 25" or "25 = 72" even though those results are also known at compile time and are likely author errors, so a WARNING should be raised when this happens, but it should still compile, and "25 = 'hello'" should be the same.

Equal(X,Y) and Equivalent(X,Y) are two of a small number of operators where X and Y really could each be any value at all and and the behaviour is well defined and reasonable.  Other operators in that small group include type membership test operators.  This reflects the real world that a type system is meant to be modelling.  A person can pick up two arbitrary objects from a bag or off the ground and can evaluate "are these the same?" regardless of how the objects otherwise are classified.  The logic is quite simple, and is unlike most other operators.

I am NOT proposing that most other "comparison" operators be defined to work for {Any,Any}; operators like less-than and greater-than etc should still only be defined within more specific types and not in general because it is more clear in the general case that there is no native way to order some things relative to each other, and if the author wanted to impose such an ordering they would do so pairwise declaring a function for each type that might be used together.

It is true that consistency across the language is important and so behaviour of Equal and Equivalent would or should correspond to behaviour of operators for generic collections such as the Contains() operator.

```
define a1: List<Any>{5,'hello',true}
define a2: a1 contains 5
define a3: a1 contains 'hello'
define a4: a1 contains 7
```

Here we have a List with a mixture of member types and the above code compiles and executes, resulting in true for a2 and a3 and doesn't result in true for a4.  Asking if X = Y is logically equivalent to asking if a set containing only X contains Y; currently the former doesn't compile while the latter does.

The ability to write generic functions is important even for CQL as that makes it easier to for library authors to factor their logic into reusable pieces.  There may be generic operations that a library author wants to perform but for which there is no corresponding operator built-in to the language.  The fact that CQL already has functions and a type hierarchy and generic types and operators built-in implies that these are desired features.  Extensibility is important as it means the core language doesn't have to be everything to everyone and authors can define their own functions to cover gaps without those having to be in the core.

CQL already defines the Any type for a reason.  It would be a shame to go to the trouble to include that type without actually exploiting its existence in reasonably useful ways.

**Comment by duncand** [comment]
For those who find difficulty with my argument and like = Equal() to be strict to particular subtypes of Any to help catch user mistakes, I propose a middle ground that might be better for everyone.

I propose that only ~ Equivalent() is altered the way I propose and that Equal() is left alone.

The way I see it, the use cases wanting to easily compare that any 2 given things are the same value would most likely be wanting to use Equivalent() anyway.

Equivalence() is already the "fundamental sameness test" whose semantics are what all the collection type operators already use such as Contains() or Includes() or Intersect() and so on.  And operators like those are what I am trying to get better semantics matches with, and they already allow collections of mixed types.

Whereas, Equal() is more of a friendly wrapper for Equivalence() for the most common use cases that fits along with less-than or basic normal author logic rather than using Equivalence() directly.

So I modify my proposal such that only Equivalent() and NotEquivalent() are modified to successfully translate when given {Any,Any} such as Equivalent(42,'hello') and Equal() can continue to die with such.

**Comment by brynrhodes** [comment]
In addressing #287, it's clear that we already support this in the translator, at least implicitly, with expressions like:

    { 1, 2, 3 } = { 'a', 'b', 'c' }

So I made several improvements to the translator overall to address this and that issue:

1. Added type constraints to the generic instantiation mechanism
2. Added common super type determination to generic instantiation
3. Changed Equal and Equivalent signatures to (Any, Any) rather than generics

These changes resolve the inconsistency, which was the most significant issue, but the changes also mean that `5 = 'a'` now translates without errors. Although I think there was value in the compile-time detection of a comparison that would always be false, I think the consistency of definitions between simple values and lists outweighs it.

Included in the [1.3 CQL Ballot](https://github.com/cqframework/clinical_quality_language/releases/tag/v1.3.0-SNAPSHOT) release.
